treon/treon.py

```diff
@@ -107,24 +107,40 @@ def print_test_collection(notebooks):
 
 
 def get_notebooks_to_test(args):
-    path = args['PATH'] or os.getcwd()
-    result = []
-
-    if os.path.isdir(path):
-        LOG.info('Recursively scanning %s for notebooks...', path)
-        path = os.path.join(path, '')  # adds trailing slash (/) if it's missing
-        glob_path = path + '**/*.ipynb'
-        result = glob.glob(glob_path, recursive=True)
-    elif os.path.isfile(path):
-        if path.lower().endswith('.ipynb'):
-            LOG.debug('Testing notebook %s', path)
-            result = [path]
-        else:
-            sys.exit('{path} is not a Notebook'.format(path=path))
+    path = pathlib.Path(args['PATH'] or os.getcwd())
+
+    if path.is_file():
+        if path.suffix == '.ipynb':
+            return [path.as_posix()]
+        sys.exit("{path} is not a notebook".format(path=path))
+    elif path.is_dir():
+        LOG.info("Recursively scanning %s for notebooks...", path)
+        ignored = build_ignore_list(path)
+        notebooks = filter(lambda nb: nb not in ignored, path.glob('**/*.ipynb'))
+        return (nb.as_posix() for nb in notebooks)
     else:
-        sys.exit('{path} is not a valid path'.format(path=path))
-
-    if not result:
-        sys.exit('No notebooks to test in {path}'.format(path=path))
-
-    return result
+        sys.exit("{path} is not a valid path".format(path=path))
+
+    sys.exit("No notebooks to test in {path}".format(path=path))
+
+
+def build_ignore_list(search_directory):
+    """
+    Recursively searches a given directory for `.treonignore` files and
+    compiles a list of :class:`pathlib.Path` objects that should be ignored.
+    Note that this function does not return the rules specified in any
+    `.treonignore` files but instead paths to the notebooks that should
+    be ignored once the parsed rules are applied.
+
+    :param pathlib.Path search_directory: directory to search
+    """
+    globs = []
+    # For each .treonignore file in search_directory or its children...
+    for ignorefile in search_directory.glob('**/.treonignore'):
+        LOG.debug("Found ignore file %s", ignorefile.as_posix())
+        # Iterate over all of the rules in the .treonignore file...
+        with ignorefile.open() as rules:
+            for rule in rules:
+                # And find any notebooks that match those rules.
+                globs.extend(ignorefile.parent.glob(rule))
```

Comment:
- Somehow the rules are not working on my machine. I tried to debug and it looks like each rule has trailing newline character attached to it. I'm on mac if that helps. Running strip() on rules makes it work.

- rule starting with slash (/) raise `NotImplementedError("Non-relative patterns are unsupported")` even though I am using relative patterns. This is [supported](https://stackoverflow.com/a/6168768/10674324) with .gitignore.